engine/src/test/java/io/camunda/zeebe/engine/processing/incident/MultiInstanceIncidentTest.java

```diff
@@ -627,6 +628,182 @@ public void shouldResolveIncidentDueToCompletionCondition() {
+   * size.
+   */
+  @Test // regression test for #9143
+  public void shouldCreateIncidentIfOutputElementCannotBeReplacedInOutputCollection() {
+    // given
+    final var processId = "index-out-of-bounds-in-output-collection";
+    final var collectionWithThreeElements = "=[1,2,3]";
+    final var collectionWithNoElements = "=[]";
+    final var outputCollectionName = "outputItems";
+
+    final var process =
+        createProcessThatModifiesOutputCollection(
+            processId, collectionWithThreeElements, collectionWithNoElements, outputCollectionName);
+
+    ENGINE.deployment().withXmlResource(process).deploy();
+
+    // when
+    final var processInstanceKey = ENGINE.processInstance().ofBpmnProcessId(processId).create();
+
+    // then
+    final Record<IncidentRecordValue> incidentEvent =
+        RecordingExporter.incidentRecords(IncidentIntent.CREATED)
+            .withProcessInstanceKey(processInstanceKey)
+            .getFirst();
+
+    Assertions.assertThat(incidentEvent.getValue())
+        .hasErrorType(ErrorType.EXTRACT_VALUE_ERROR)
+        .hasErrorMessage(
+            "Unable to update item in output collection 'outputItems' at position 1 because the size of the collection is: 0. This happens when multiple BPMN elements write to the same variable.")
+        .hasProcessInstanceKey(processInstanceKey);
+  }
+
+  @Test
+  public void shouldResolveIncidentRaisedByOutputElementCannotBeReplacedInOutputCollection() {
+    // given
+    final var processId = "index-out-of-bounds-in-output-collection";
+    final var collectionWithOneElement = "=[1]";
+    final var collectionWithNoElements = "=[]";
+    final var outputCollectionName = "outputItems";
+
+    final var process =
+        createProcessThatModifiesOutputCollection(
+            processId, collectionWithOneElement, collectionWithNoElements, outputCollectionName);
+
+    ENGINE.deployment().withXmlResource(process).deploy();
+    final var processInstanceKey = ENGINE.processInstance().ofBpmnProcessId(processId).create();
+
+    final Record<IncidentRecordValue> incidentEvent =
+        RecordingExporter.incidentRecords(IncidentIntent.CREATED)
+            .withProcessInstanceKey(processInstanceKey)
+            .getFirst();
+    // when
+    ENGINE
+        .variables()
+        .ofScope(incidentEvent.getValue().getVariableScopeKey())
+        .withDocument(Maps.of(entry(outputCollectionName, List.of(1))))
+        .update();
+    ENGINE.incident().ofInstance(processInstanceKey).withKey(incidentEvent.getKey()).resolve();
+
+    // then the process is able to complete
+    assertThat(
+            RecordingExporter.processInstanceRecords(ProcessInstanceIntent.ELEMENT_COMPLETED)
+                .withElementType(BpmnElementType.PROCESS)
+                .withProcessInstanceKey(processInstanceKey)
+                .limitToProcessInstanceCompleted()
+                .exists())
+        .describedAs("the process has completed")
+        .isTrue();
```

Comment:
:x: In addition, we should verify that the output collection is updated with the output element after the incident has been resolved.